### DISCUSSION TITLE
Time Varying Source

### DISCUSSION THREAD

**Discussion by skilduff** [su2code/SU2#2160] [Q&A]
Hi all,

I've been trying to implement a time and spatially varying heat source to the code, but it is not behaving as expected. I've made similar changes to my code as this post: https://github.com/su2code/SU2/discussions/1904 

The main difference in my code is that my power varies with time. I obtain the time from these lines of code in flow_sources.cpp:
const auto dt = config->GetDelta_UnstTimeND();
const su2double t = config->GetTimeIter() * dt

I then calculate the source's volumetric heat flux for the current time and add it to the energy residual:
residual[nDim+1] = -Volume * Source

I have compared the values calculated in my code's source term (in W/m^3) and they match the analytical model I am comparing it against. However, when I run the simulation, the temperature increase does not match the temperature increase from the analytical model (it is off by several orders of magnitude). I think the issue may be arising from calculating and adding to the residual, but I am not so sure. 

Any help would be appreciated.

Thanks

**Comment by NAnand-TUD** [comment]
Hello, If you can point me to the branch where you have implemented this, it would be easier to give you some feedback.

**Reply by skilduff** [reply]
Hi,

Thanks for the reply. I've attached a pdf that has documented the changes I have made to the code. Let me know if there are any issues.
I am using SU2 V8.0.0.

Thanks

[SU2_source_code.pdf](https://github.com/su2code/SU2/files/13267028/SU2_source_code.pdf)

**Reply by NAnand-TUD** [reply]
Could you check your source term? You are initialing residual to iDim+1 (iDim max being nDim). However, your residual value is being set at residual [nDim+1].

`residual[nDim+1] = -Volume * Spe;`

This may be the source of the problem. But definitely, on the first look, this looks like a problem.

Thanks,
Nitish

**Reply by skilduff** [reply]
I'm not sure I quite understand. Is residual[nDim+1] not the energy contribution to the residual? I based this off of CSourceBodyForce, so I assumed this was correct.

My understanding was that for a 2D compressible flow:
residual[0] - continuity contribution 
residual[1] - x-momentum contribution
residual[2] - y-momentum contribution
residual[3] - energy contribution

Is this correct?

**Reply by pcarruscag** [reply]
That's correct, have you printed some values from the code to see if they match expectations?
Is your simulation non-dimensional? Perhaps you are missing some dimensional factors.

**Reply by NAnand-TUD** [reply]
Right, got it. Disregard my comment @skilduff .

**Reply by skilduff** [reply]
My simulation is dimensional. I have printed Spe (the source term in W/m^3) and compared it to my analytical model; the results are the same. However, my analytical model also predicts an increase in temperature of ~250 K, whereas my simulation shows the temperature does not increase even 1 K.

**Reply by pcarruscag** [reply]
Have you done some simple verification? For example an adiabatic channel flow with your source term and verify that enthalpy out = enthalpy in + integrated source

**Reply by skilduff** [reply]
Yes, I am trying to validate the model by simulating an essentially 0 velocity flow in a 1m x 1m square. 

I am using this equation:

$$\frac{d}{dt}\rho e =  S$$

I am also assuming constant density and a calorically perfect gas: $$e = C_vT$$
